Adding a Cache Driver

jamesstocktonj1 opened the pull request:
# Overview
This feature adds a Cache driver which provides a simple look-aside caching pattern. It satisfies the Database interface and takes in two other `hord.Database` in it's configuration. The following is an example is a setup using a Cassandra database with a Redis cache:

```
// Handle Cassandra connection
var cdb hord.Database
cdb, err := cassandra.Dial(cassandra.Config{})
if err != nil {
	// Handle connection error
}

// Handle Redis connection
var rdb hord.Database
rdb, err = redis.Dial(redis.Config{})
if err != nil {
	// Handle connection error
}

var db hord.Database
db, err := cache.Dial(cache.Config{
	Database: cdb,
	Cache: rdb,
})
if err != nil {
	// Handle connection error
}
```
The following is a quick benchmark of the setup found above:
```
Testing with 10000 keys
Set: 14.020334231s
Get: 1.070310995s
Get (Uncached): 12.615694324s
```
In conclusion, this Cache driver provides a boost in performance within the same Database interface.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Introduced a new package `cache` with functions for database operations and a `Cache` struct for managing data.
	- Added a `lookaside` package for a database driver to handle cache operations.
- **Tests**
	- Included test functions for the `cache` and `lookaside` packages to validate database interactions and cache functionality.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jamesstocktonj1 committed:
Add look-aside cache implementation

jamesstocktonj1 committed:
Add nil checks

jamesstocktonj1 committed:
Add cache common tests

jamesstocktonj1 committed:
Add cache driver to tests workflow

jamesstocktonj1 committed:
Fix linting

jamesstocktonj1 committed:
Add cache update test coverage

jamesstocktonj1 committed:
Fix linting

madflojo commented:
This looks solid, and I like the way you implemented it. I am debating if we should create a new directory alongside `drivers/` called `caching/` and put this as a `lookaside/` package within caching. That way, other caching strategies can also be contributed, and what they do will be evident.

Thoughts?

madflojo commented on drivers/cache/cache.go:
```diff
@@ -0,0 +1,244 @@
+	// Update the cache
+	err = db.cache.Set(key, data)
+	if err != nil {
+		return data, err
```
I wonder if wrapping the original driver error with an `ErrCacheError` or an `ErrDatabaseError` error type makes sense.

Today drivers return these errors, but wrapping them might give users more ideas on where the error was.
https://github.com/madflojo/hord/blob/main/hord.go#L100-L106

jamesstocktonj1 replied:
I was debating whether to return the error at all considering the read has been successful and the cache write-back could be considered non-essential.

jamesstocktonj1 replied:
Have wrapped in a ErrCacheError error.

madflojo commented on drivers/cache/cache.go:
```diff
@@ -0,0 +1,244 @@
+	return nil
+}
+
+// Get will get the data from the database. It uses a look-aside pattern to store the data in the cache if it is not already there.
```
```suggestion
// Get will get the data from the cache database. If not found, it uses a look-aside pattern to fetch from the data database and store the data in the cache.
```

madflojo commented on drivers/cache/cache.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// Cache is used to store data in a look-aside caching pattern. It also satisfies the Hord database interface.
+type Cache struct {
```
Non-blocker: It may be useful to give users methods to retrieve the `data` and `cache`. In theory, they should have access to it, but if this struct instance is passed down several packages, they may not, but they may find the need to access the cache or database directly outside of the caching pattern.

It's a thought, but I wouldn't hold up the PR for it.

jamesstocktonj1 committed:
Move lookaside

jamesstocktonj1 committed:
Add cache write-back error wrapper

jamesstocktonj1 committed:
Fix redundant nil check

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jamesstocktonj1 commented:
> This looks solid, and I like the way you implemented it. I am debating if we should create a new directory alongside `drivers/` called `caching/` and put this as a `lookaside/` package within caching. That way, other caching strategies can also be contributed, and what they do will be evident.
> 
> Thoughts?

I have moved the implementation to the `lookaside/` subfolder in cache and added a `CacheType` to the `Config` struct in cache so more caching strategies can be implemented.

As for the project structure, simplicity would say to have all the packages that satisfy the `hord.Database` interface in the same directory. On the other hand, the cache package isn't really considered a driver but more of a layer of abstraction so would make sense to separate it.

I'm easy either way so will leave the decision to you.

jamesstocktonj1 committed:
Add default to lookaside

jamesstocktonj1 committed:
Revert default lookaside + Add GetCache and GetDatabase functions

jamesstocktonj1 committed:
Fix nil Dial return

jamesstocktonj1 committed:
Increase test coverage

madflojo commented:
> As for the project structure, simplicity would say to have all the packages that satisfy the `hord.Database` interface in the same directory. On the other hand, the cache package isn't really considered a driver but more of a layer of abstraction so would make sense to separate it.
> 
> I'm easy either way so will leave the decision to you.

Yeah, I think moving the cache directory up one layer makes the most sense to me. Like you said, it's not really a driver but more of an abstraction.

Thank you again for this contribution!

jamesstocktonj1 commented:
You're welcome, it's a pleasure to contribute. I look forward to seeing where this project goes.
